drivers/crc/crc_renesas_ra.c

```diff
@@ -0,0 +1,282 @@
+}
+
+static int crc_renesas_ra_begin(const struct device *dev, struct crc_ctx *ctx)
+{
+	int ret;
+
+	crc_lock(dev);
+
+	ret = crc_set_config(dev, ctx);
+	if (ret != 0) {
+		crc_unlock(dev);
+		return ret;
+	}
+
+	ctx->state = CRC_STATE_IN_PROGRESS;
+
+	return 0;
+}
+
+static int crc_renesas_ra_update(const struct device *dev, struct crc_ctx *ctx, const void *buffer,
+				 size_t bufsize)
+{
+	struct crc_renesas_ra_data *data = dev->data;
+	fsp_err_t err;
+	uint32_t init_val;
+
+	/* Ensure CRC calculation has been initialized by crc_begin() */
+	if (ctx->state == CRC_STATE_IDLE) {
+		return -EINVAL;
+	}
+
+	switch (ctx->type) {
+	case CRC8:
+		__fallthrough;
+	case CRC16:
+		__fallthrough;
+	case CRC16_CCITT: {
+		if (ctx->type == CRC8) {
+			init_val = (data->flag_crc_updated ? ctx->result : ctx->seed) & 0xFF;
+		} else {
+			init_val = (data->flag_crc_updated ? ctx->result : ctx->seed) & 0xFFFF;
+		}
+
+		data->input_data.num_bytes = bufsize;
+		data->input_data.crc_seed = init_val;
+		data->input_data.p_input_buffer = (uint8_t *)buffer;
+
+		err = R_CRC_Calculate(&data->ctrl, &data->input_data, &ctx->result);
+		if (err != FSP_SUCCESS) {
+			ctx->state = CRC_STATE_IDLE;
+			crc_unlock(dev);
+			return -EINVAL;
+		}
+		break;
+	}
+	default: {
+		init_val = (data->flag_crc_updated ? ctx->result : ctx->seed) & 0xFFFFFFFF;
+
+		if ((bufsize % 4) != 0) {
+			ctx->state = CRC_STATE_IDLE;
+			crc_unlock(dev);
+			return -ENOTSUP;
+		}
+
+		data->input_data.num_bytes = bufsize;
+		data->input_data.crc_seed = init_val;
+		data->input_data.p_input_buffer = (uint8_t *)buffer;
+
+		err = R_CRC_Calculate(&data->ctrl, &data->input_data, &ctx->result);
+		if (err != FSP_SUCCESS) {
+			ctx->state = CRC_STATE_IDLE;
+			crc_unlock(dev);
+			return -EINVAL;
+		}
+
+		if (ctx->type == CRC32_IEEE) {
+			ctx->result = (crc_result_t)~ctx->result;
+		}
+		break;
+	}
+	}
+
+	data->flag_crc_updated = true;
+
+	return 0;
+}
+
+static int crc_renesas_ra_finish(const struct device *dev, struct crc_ctx *ctx)
+{
+	struct crc_renesas_ra_data *data = dev->data;
+
+	if (ctx->state == CRC_STATE_IDLE) {
+		return -EINVAL;
+	}
+
+	ctx->state = CRC_STATE_IDLE;
+
+	data->flag_crc_updated = false;
+
+	crc_unlock(dev);
+
+	return 0;
+}
+
+static int crc_ra_init(const struct device *dev)
+{
+	int ret;
+	fsp_err_t err;
+	const struct crc_renesas_ra_cfg *cfg = dev->config;
+	struct crc_renesas_ra_data *data = dev->data;
+	crc_cfg_t const *const crc_cfg = &data->crc_config;
+
+	if (!device_is_ready(cfg->clock_dev)) {
+		LOG_ERR("CRC: Clock control device not ready");
+		return -ENODEV;
+	}
+
+	ret = clock_control_on(cfg->clock_dev, (clock_control_subsys_t)&cfg->clock_id);
+	if (ret < 0) {
+		LOG_ERR("CRC: Clock control device could not initialize");
+		return -EIO;
+	}
+
+	err = R_CRC_Open(&data->ctrl, crc_cfg);
+	if (err != FSP_SUCCESS) {
+		return -EINVAL;
+	}
+
+	k_sem_init(&data->sem, 1, 1);
+
+	return 0;
+}
+
+static const struct crc_driver_api crc_renesas_ra_driver_api = {
```

Comment:
Use the device API macro to place this in a dedicated linker section.

```suggestion
static DEVICE_API(crc, crc_renesas_ra_driver_api) = {
```

Reply:
I fixed it